### PR TITLE
core/rawdb: avoid unnecessary receipt processing for log filtering

### PR DESCRIPTION
**Pull request by s1na**
This is a successor for #23058. In this PR, the new lightweight `receiptLogs` type has been instead moved to rawdb and made private.

### CODE REVIEW

**Review by OhArthits — APPROVED** [review]

**Comment by s1na** [comment]
To see how these types are used refer here: https://github.com/s1na/go-ethereum/pull/5

**Review by holiman — APPROVED** [review]
LGTM

**Review by OhArthits — APPROVED** [review]

**Comment by s1na** [comment]
To estimate how effective the triage suggestion https://github.com/ethereum/go-ethereum/pull/23058#discussion_r656024846 would be I added 2 metrics in https://github.com/ethereum/go-ethereum/pull/23520 which show us how often we lookup a block with zero matching logs and it seems to happen in ~1% of cases. That means we can avoid reading block body from db in 1% of cases by filtering deeper in the stack.

**Comment by s1na** [comment]
I did a hacky implementation of the optimization in https://github.com/s1na/go-ethereum/commit/7fd9ac91a394d31c03c4c4498cdb1d180d22acdf and measured the time for a few contracts/events (see [script](https://gist.github.com/s1na/f85df89173bac8c7a66077cddfec1c7c)). The script sends many subsequent `getLogs` requests, each for a range of 2000 blocks.

Screenshot 2021-09-06 at 18 25 45 cropped

The right-side chart is the mean time for responding to a request and left-side is the mean time for processing one matching block within a request. I stopped the node twice and changed branches:
- Last numbers are from baseline (master)
- Middle ones is this PR
- In the first one we avoid reading the block body from db when it has zero matching logs

File: eth/api_backend.go [hunk 1]
```diff
@@ -181,13 +181,14 @@ func (b *EthAPIBackend) GetReceipts(ctx context.Context, hash common.Hash) (type
...
+	db := b.eth.ChainDb()
+	number := rawdb.ReadHeaderNumber(db, hash)
+	if number == nil {
+		return nil, errors.New("failed to get block number from hash")
 	}
-	logs := make([][]*types.Log, len(receipts))
-	for i, receipt := range receipts {
-		logs[i] = receipt.Logs
+	logs := rawdb.ReadLogs(db, hash, *number)
+	if logs == nil {
```
Comment (MariusVanDerWijden) [inline eth/api_backend.go:190]: I would prefer ReadLogs to return an error instead of relying on logs to be non nil here

**Comment by s1na** [comment]
Added a benchmark to compare decoding a stored receipt the normal way (via `ReceiptForStorage`) against the new type `rlpLogs`. Results from my own machine:

```
BenchmarkDecodeRLPLogs
BenchmarkDecodeRLPLogs/ReceiptForStorage
BenchmarkDecodeRLPLogs/ReceiptForStorage-8              1000000000               0.00242 ns/op
BenchmarkDecodeRLPLogs/ReceiptForStorage-8              1000000000               0.00257 ns/op
BenchmarkDecodeRLPLogs/ReceiptForStorage-8              1000000000               0.00273 ns/op
BenchmarkDecodeRLPLogs/ReceiptForStorage-8              1000000000               0.00268 ns/op
BenchmarkDecodeRLPLogs/ReceiptForStorage-8              1000000000               0.00263 ns/op
BenchmarkDecodeRLPLogs/rlpLogs
BenchmarkDecodeRLPLogs/rlpLogs-8                        1000000000               0.000877 ns/op
BenchmarkDecodeRLPLogs/rlpLogs-8                        1000000000               0.000989 ns/op
BenchmarkDecodeRLPLogs/rlpLogs-8                        1000000000               0.000871 ns/op
BenchmarkDecodeRLPLogs/rlpLogs-8                        1000000000               0.000871 ns/op
BenchmarkDecodeRLPLogs/rlpLogs-8                        1000000000               0.000926 ns/op
```

**Comment by s1na** [comment]
Thanks for the catch. Here are the updated result:

```
BenchmarkDecodeRLPLogs
BenchmarkDecodeRLPLogs/ReceiptForStorage
BenchmarkDecodeRLPLogs/ReceiptForStorage-8                   452           2589949 ns/op          660146 B/op       8021 allocs/op
BenchmarkDecodeRLPLogs/ReceiptForStorage-8                   452           2593112 ns/op          660132 B/op       8020 allocs/op
BenchmarkDecodeRLPLogs/ReceiptForStorage-8                   453           2568350 ns/op          660157 B/op       8021 allocs/op
BenchmarkDecodeRLPLogs/ReceiptForStorage-8                   459           2568151 ns/op          660126 B/op       8020 allocs/op
BenchmarkDecodeRLPLogs/ReceiptForStorage-8                   453           2599324 ns/op          660153 B/op       8021 allocs/op
BenchmarkDecodeRLPLogs/rlpLogs
BenchmarkDecodeRLPLogs/rlpLogs-8                            1237            970215 ns/op          497811 B/op       5870 allocs/op
BenchmarkDecodeRLPLogs/rlpLogs-8                            1215            966026 ns/op          497810 B/op       5870 allocs/op
BenchmarkDecodeRLPLogs/rlpLogs-8                            1234           1019230 ns/op          497813 B/op       5870 allocs/op
BenchmarkDecodeRLPLogs/rlpLogs-8                            1248            974802 ns/op          497814 B/op       5870 allocs/op
BenchmarkDecodeRLPLogs/rlpLogs-8                            1196            965926 ns/op          497813 B/op       5870 allocs/op
```

**Comment by s1na** [comment]
- I re-ran the benchmarks with geth disconnected from the network. This clearly removes the creep-up effect.
- Also did a few runs of each variant in semi-random order to make the distinction more visible.
- Finally, the disk read meter shows that optim2 is in fact reading less data on the margin. Optim2 is around 6.48 and optim1 around 6.77
    - The chaindata data read meter has more variance, not sure why

Screenshot 2021-09-08 at 10 47 21

**Review by holiman — APPROVED** [review]
LGTM 
"faster is better"

**Comment by holiman** [comment]
Oh no needs a rebase

File: eth/api_backend.go (hunk 1 above)
Comment (s1na) [reply eth/api_backend.go:190]: I hear you. just didn't return an error to be consistent with the rest of the accessors

**Comment by s1na** [comment]
@holiman I rebased. Just a reminder that this PR only has Optim1. Didn't want to make it too big by including Optim2 as well

**Review by fjl — APPROVED** [review]

**Pull request merged**